File: LeetCode/85/main.cpp
#include <stack>

class Solution {
public:

    int get_height(int r, int c, int m, vector<vector<char>>& mat) {
        int res = 0;
        for(int j = c; j < m; j++) {
            if(mat[r][j] == '0') {
                return res;
            }
            res += 1;
        }
        return res;
    }

    int maximalRectangle1D(vector<int> arr) {
        int n = arr.size();
        int lefts[n], rights[n];
        int ans = 0;
        stack<pair<int, int>> st;

        for(int i = 0; i < n; i++){
            while(!st.empty() && arr[i] <= st.top().first) {
                st.pop();
            }
            if(st.empty()) {
                lefts[i] = 0;
            } else {
                lefts[i] = st.top().second + 1;
            }
            st.push({arr[i], i});
        }

        while(!st.empty()) {
            st.pop();
        }

        for(int i = n - 1; i >= 0; i--){
            while(!st.empty() && arr[i] <= st.top().first) {
                st.pop();
            }
            if(st.empty()) {
                rights[i] = (n - 1);
            } else {
                rights[i] = st.top().second - 1;
            }
            st.push({arr[i], i});
        }

        for(int i = 0; i < n; i++) {
            ans = max(ans, arr[i] * (rights[i] - lefts[i] + 1));
        }

        return ans;

    }

    int maximalRectangle(vector<vector<char>>& mat) {
        int n = mat.size();
        int m = mat[0].size();
        int ans = 0;

        vector<vector<int>> arrs;
        vector<int> hs;

        for(int j = 0; j < m; j++) {
            hs.clear();
            for(int i = 0; i < n; i++) {
                int h = get_height(i, j, m, mat);
                hs.emplace_back(h);
            }
            arrs.emplace_back(hs);
        }

        for(auto& arr: arrs) {
            int res = maximalRectangle1D(arr);
            ans = max(ans, res);
        }
        
        return ans;
        
    }
};

File: LeetCode/85/main.py
from collections import deque


class Solution:
    def maximalRectangle(self, mat: List[List[str]]) -> int:
        n = len(mat)  # no. of rows
        m = len(mat[0])  # no. of cols
        arr = []
        ans = 0

        def get_height(r, c):
            row = list(map(int, mat[r][c:]))
            res = 0
            for x in row:
                if x == 0:
                    break
                res += 1
            return res

        for j in range(m):
            hs = []
            for i in range(n):
                hs.append(get_height(i, j))
            arr.append(hs)

        def largest_rectangle_area(a):
            n = len(a)
            ans = 0
            st = deque()
            lefts = [0] * n
            rights = [0] * n

            for i in range(n):
                while len(st) > 0 and a[i] <= st[-1][0]:
                    st.pop()
                lefts[i] = 0 if len(st) == 0 else st[-1][1] + 1
                st.append((a[i], i))

            while len(st) > 0:
                st.pop()

            for i in range(n - 1, -1, -1):
                while len(st) > 0 and a[i] <= st[-1][0]:
                    st.pop()

                rights[i] = (n - 1) if len(st) == 0 else st[-1][1] - 1
                st.append((a[i], i))

            for i in range(n):
                ans = max(ans, a[i] * (rights[i] - lefts[i] + 1))

            return ans

        for i, a in enumerate(arr):
            ans = max(ans, largest_rectangle_area(a))

        return ans
